File: src/shared/PlantBasedPizza.Events/EventBridgeEventPublisher.cs
using Amazon.EventBridge;
using Amazon.EventBridge.Model;
using CloudNative.CloudEvents;
using CloudNative.CloudEvents.SystemTextJson;
using Datadog.Trace;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using System.Text.Json;

namespace PlantBasedPizza.Events;

public class EventBridgeEventPublisher : IEventPublisher
{
    private readonly AmazonEventBridgeClient _eventBridgeClient;
    private readonly EventBridgeSettings _settings;
    private readonly ILogger<EventBridgeEventPublisher> _logger;

    public EventBridgeEventPublisher(AmazonEventBridgeClient eventBridgeClient, IOptions<EventBridgeSettings> settings)
    {
        _eventBridgeClient = eventBridgeClient;
        _settings = settings.Value;
    }

    public async Task Publish(IntegrationEvent evt)
    {
        var eventType = $"{evt.EventName}.{evt.EventVersion}";
        
        var eventId = Guid.NewGuid()
            .ToString();

        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.eventId", eventId);
        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.eventType", eventType);
        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.eventName", evt.EventName);
        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.eventVersion", evt.EventVersion);
        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.eventSource", evt.Source.ToString());
        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.busName", _settings.BusName);

        var evtWrapper = new CloudEvent
        {
            Type = eventType,
            Source = evt.Source,
            Time = DateTimeOffset.Now,
            DataContentType = "application/json",
            Id = eventId,
            Data = evt,
        };
        
        if (Tracer.Instance.ActiveScope?.Span != null)
        {
            var serializedHeaders = "";

            try
            {
                Console.WriteLine("Injecting headers");
                var spanInjector = new SpanContextInjector();
                var headers = new Dictionary<string, string>();
            
                spanInjector.Inject("datadog", (s, s1, arg3) =>
                {
                    headers.Add(s1, arg3);
                }, Tracer.Instance.ActiveScope.Span.Context);

                serializedHeaders = JsonSerializer.Serialize(headers);
                Console.WriteLine(serializedHeaders);
            }
            catch (Exception e)
            {
                Console.WriteLine("Error manually injecting headers");
                Console.WriteLine(e);
            }
            
            evtWrapper.SetAttributeFromString("ddtraceid", Tracer.Instance.ActiveScope.Span.TraceId.ToString());
            evtWrapper.SetAttributeFromString("ddspanid", Tracer.Instance.ActiveScope.Span.SpanId.ToString());
            evtWrapper.SetAttributeFromString("tracedata", serializedHeaders);
        }

        var evtFormatter = new JsonEventFormatter();

        var json = evtFormatter.ConvertToJsonElement(evtWrapper).ToString();

        var publishResponse = await _eventBridgeClient.PutEventsAsync(new PutEventsRequest()
        {
            Entries = new List<PutEventsRequestEntry>(1)
            {
                new()
                {
                    Source = evt.Source.ToString(),
                    EventBusName = _settings.BusName,
                    Detail = json,
                    DetailType = eventType
                }
            }
        });

        Tracer.Instance.ActiveScope?.Span.SetTag("messaging.failedEvents", publishResponse.FailedEntryCount);

        if (publishResponse.FailedEntryCount > 0)
        {
            foreach (var failedEvent in publishResponse.Entries.Where(p => string.IsNullOrEmpty(p.EventId)))
            {
                this._logger.LogError("Event publish failed with {ErrorCode} and {ErrorMessage}", failedEvent.ErrorCode, failedEvent.ErrorMessage);
            }
        }
    }
}

File: src/PlantBasedPizza.Kitchen.Node/src/adapters/eventBridgeEventPublisher.ts
import { CloudEventV1 } from "cloudevents";
import { IKitchenEventPublisher } from "../entities/kitchenEventPublisher";
import { KitchenOrderConfirmedEventV1 } from "../events/kitchenOrderConfirmedV1Event";
import { EventBridgeClient, PutEventsCommand } from "@aws-sdk/client-eventbridge";
import { tracer } from "dd-trace";
import { OrderBakedEventV1 } from "../events/orderBakedEventV1";
import { OrderPrepCompleteEventV1 } from "../events/orderPrepCompleteEventV1";
import { OrderPreparingEventV1 } from "../events/orderPreparingEventV1";
import { OrderQualityCheckedEventV1 } from "../events/orderQualityCheckedEventV1";
import { v4 as uuidv4 } from 'uuid';

const { FORMAT_HTTP_HEADERS } = require('opentracing')
const { getTraceHeaders } = require("datadog-lambda-js");

export class EventBridgeEventPublisher implements IKitchenEventPublisher {
  private eventBridgeClient: EventBridgeClient;

  constructor(eventBridgeClient: EventBridgeClient) {
    this.eventBridgeClient = eventBridgeClient;
  }
  async publishOrderBakedEventV1(evt: OrderBakedEventV1): Promise<void> {
    await this.publish("kitchen.orderBaked.v1", evt);
  }
  async publishOrderPrepCompleteEventV1(evt: OrderPrepCompleteEventV1): Promise<void> {
    await this.publish("kitchen.orderPrepComplete.v1", evt);
  }
  async publishOrderPreparingEventV1(evt: OrderPreparingEventV1): Promise<void> {
    await this.publish("kitchen.orderPreparing.v1", evt);
  }
  async publishOrderQualityCheckedEventV1(evt: OrderQualityCheckedEventV1): Promise<void> {
    await this.publish("kitchen.qualityChecked.v1", evt);
  }

  async publishKitchenOrderConfirmedEventV1(evt: KitchenOrderConfirmedEventV1): Promise<void> {
    await this.publish("kitchen.orderConfirmed.v1", evt);
  }

  async publish<T>(evtType: string, evtData: T) {
    const currentSpan = tracer.scope().active();

    const span= tracer.scope().active()!;
    const headers = {};

    tracer.inject(span, FORMAT_HTTP_HEADERS, headers);

    const ce: CloudEventV1<T> = {
      specversion: "1.0",
      source: "https://kitchen.plantbasedpizza/",
      type: evtType,
      id: uuidv4(),
      time: new Date().toISOString(),
      datacontenttype: "application/json",
      data: evtData,
      ddtraceid: currentSpan?.context().toTraceId(),
      ddspanid: currentSpan?.context().toSpanId(),
      traceparent: currentSpan?.context().toTraceparent(),
    };

    currentSpan?.addTags({
      "messaging.eventId": ce.id,
      "messaging.eventType": ce.type,
      "messaging.eventSource": ce.source,
      "messaging.busName": process.env.BUS_NAME,
    });

    const command = new PutEventsCommand({
      Entries: [
        {
          Detail: JSON.stringify(ce),
          DetailType: ce.type,
          EventBusName: process.env.BUS_NAME,
          Source: ce.source,
          Time: new Date(),
        },
      ],
    });

    const ebResult = await this.eventBridgeClient.send(command);

    currentSpan?.addTags({"messaging.failedEvents": ebResult.FailedEntryCount})
  }
}
